sdk/helper/storagepacker/partition.go

```diff
@@ -0,0 +1,115 @@
+package storagepacker
+
+import (
+	"fmt"
+	"github.com/hashicorp/vault/sdk/helper/locksutil"
+)
+
+// Requests for multi-item put or get, partitioned by bucket.
+// This lets us operate bucket-by-bucket.
+type partitionedRequests struct {
+	Bucket   *LockedBucket
+	Requests []*itemRequest
+}
+
+// partitionRequests takes a list of requests and partition them by which
+// bucket the Key belongs to.
+func (s *StoragePackerV2) partitionRequests(unsortedRequests []*itemRequest) ([]*partitionedRequests, error) {
+	sortedRequests := sortRequests(unsortedRequests)
+
+	partition := make([]*partitionedRequests, 0)
+	var lastBucket *partitionedRequests
+
+	// The items requests are sorted by key, so the same buckets end
+	// up together. But, the radix tree doesn't have a way to take
+	// advantage of that to do a single pass.
```

Comment:
Doesn't this mean then that we shouldn't sort it? We're not benefiting but we're taking the efficiency hit.

Reply:
While sorting doesn't change the number of radix tree lookups we have to do, it does mean that we can produce the buckets in-order, and I do take advantage of that.

I think now it is possible to make the partitioning somewhat more efficient but it would be a little more complex.  Because we always fully shard a bucket, once we do a lookup, we can just compare whether the same-length prefix of the next key is the same.  If the radix tree were less regular that wouldn't work.

So, for example, if our keys are 
AABB
AABC
AACD
ABFF
then if radix lookup returns AA/B we can take the first two keys and do our next lookup of AACD.  If radix lookup returns just AA we can take the first three keys. and do the next lookup on ABFF.

Reply:
Oh, I forgot I did mention that alternative in the comment.